The Python SDK community edition

sebastian-quintero opened the pull request:
# Description

This PR finishes the community edition with the local package.

sebastian-quintero committed:
Add new_local_run

sebastian-quintero committed:
Methods for getting local results

sebastian-quintero committed:
Methods for syncing local runs to Cloud

sebastian-quintero committed:
Fixes type hints in TrackedRun

Corrects type hints for optional
attributes in TrackedRun class.

sebastian-quintero committed:
Adds visualization features to local experience

sebastian-quintero committed:
Uses `python` to run executor

Uses `python` executable instead of
`sys.executable` to run the executor.

sebastian-quintero committed:
Move all the local stuff to the local package

sebastian-quintero committed:
Refactors local runs

Streamlines the local run process.

Fixes application init, logs and sync.

sebastian-quintero commented on docs/nextmv/reference/cloud/safe.md:
```diff
@@ -0,0 +1,5 @@
+# Safe Module
+
+This section documents the sagfe components of the Nextmv Cloud API.
```
Done

sebastian-quintero commented on nextmv/nextmv/cloud/application.py:
```diff
@@ -3594,6 +3705,74 @@ def __upload_url_required(
+
+        options_dict = {}
+        if isinstance(input, Input) and input.options is not None:
+            options_dict = input.options.to_dict_cloud()
```
Thanks

sebastian-quintero committed:
Fixes typo and handles options

Fixes a typo in documentation.

Handles serialization of options.

sebastian-quintero committed:
Merge branch 'develop' into feature/eng-6317-add-ability-to-execute-a-local-run

sebastian-quintero committed:
Merge branch 'feature/eng-6317-add-ability-to-execute-a-local-run' into feature/eng-6324-retrieve-local-run

sebastian-quintero committed:
Merge branch 'feature/eng-6324-retrieve-local-run' into feature/eng-6323-sync-local-application-runs-to-remote-application

sebastian-quintero committed:
Merge branch 'feature/eng-6323-sync-local-application-runs-to-remote-application' into feature/eng-6326-local-visualization

sebastian-quintero committed:
Merge branch 'feature/eng-6326-local-visualization' into feature/change-local-namespace

sebastian-quintero committed:
Fixes typo in docs

Corrects a typo in the documentation.

Removes obsolete input options logic.

sebastian-quintero committed:
Processed copilot suggestions

sebastian-quintero committed:
Standardizes file structure

Uses constants to define file paths,
improving code maintainability.

sebastian-quintero committed:
Merge pull request #137 from nextmv-io/feature/change-local-namespace

sebastian-quintero committed:
Merge pull request #135 from nextmv-io/feature/eng-6326-local-visualization

sebastian-quintero committed:
Merge pull request #134 from nextmv-io/feature/eng-6323-sync-local-application-runs-to-remote-application

sebastian-quintero committed:
Merge pull request #133 from nextmv-io/feature/eng-6324-retrieve-local-run

merschformann commented on nextmv/nextmv/cloud/application.py:
```diff
@@ -2677,13 +2387,26 @@ def track_run(self, tracked_run: TrackedRun, instance_id: Optional[str] = None)
+        there are important considerations to take into account when using this
+        method. For example, if you intend to upload JSON input/output, use the
+        `input`/`output` attributes of the `TrackedRun` class. On the other
+        hand, if you intend to track files-based input/output, use the
+        `input_dir_path`/`output_dir_path` attributes of the `TrackedRun`
+        class.
```
I don't think non-json tracked runs are supported yet. It might be on the horizon though. Is this aligned with a future extension?

sebastian-quintero replied:
Baking in support to be future-proof.

sebastian-quintero committed:
Merge branch 'develop' into feature/eng-6317-add-ability-to-execute-a-local-run

sebastian-quintero committed:
Uses default polling options

Uses the configured default
polling options in several
functions.

merschformann commented on nextmv/nextmv/local/application.py:
```diff
@@ -0,0 +1,1145 @@
+
+        # For files-based runs, the result files are written by default to ".".
+        # Avoid this using a dedicated temp dir.
+        run_result = self.run_result(run_id, output_dir_path=temp_dir)
```
How do we make sure that the run-ids do not collide? The run-id is specified by the user, right? :thinking:

sebastian-quintero replied:
The run ID is not user generated

merschformann commented on nextmv/nextmv/local/application.py:
```diff
@@ -0,0 +1,1145 @@
+                format_output=run_result.metadata.format.format_output,
+            ),
+        )
+        tracked_id = target.track_run(
```
When syncing a run with problematic input (binary files, etc.) or output (too large, etc.) the sync will simply fail, right? Will it try to re-sync and fail every time the user tries to sync all runs? :thinking:

sebastian-quintero replied:
Correct, if syncing a run is problematic from `cloud` side, then syncing that run fails.

merschformann commented on nextmv/nextmv/local/executor.py:
```diff
@@ -0,0 +1,565 @@
+                json.dump(info, f, indent=2)
+                f.truncate()
+
+            # Start a Python subprocess to execute the entrypoint. For now, we are
+            # supporting a Python-first experience, so we are not summoning
+            # applications that are not Python-based.
```
Thanks for clarifying, I was wondering about that. :relaxed:

merschformann commented on nextmv/nextmv/local/executor.py:
```diff
@@ -0,0 +1,565 @@
+        # place to work from, and be cleaned up afterwards.
+        with tempfile.TemporaryDirectory() as temp_dir:
+            temp_src = os.path.join(temp_dir, "src")
+            shutil.copytree(src, temp_src, ignore=shutil.ignore_patterns(".nextmv"))
```
We are not copying / handling the requirements from the manifest / requirements.txt file, right? :thinking: 
Is this a plan in the future, or, do we simply use the local installation for the local experience?

sebastian-quintero replied:
We use the local installation given by `sys.executable`, so we don't care much about the requirements at the moment.

merschformann commented on nextmv/nextmv/local/executor.py:
```diff
@@ -0,0 +1,565 @@
+        The result of the subprocess run.
+    """
+
+    logs_dir = os.path.join(run_dir, LOGS_KEY)
+    os.makedirs(logs_dir, exist_ok=True)
+    with open(os.path.join(logs_dir, LOGS_FILE), "w") as f:
+        f.write(result.stderr)
```
I think we need a bit more complex stdout/stderr handling logic. For `multi-file` we consider stdout+stderr as the log-stream. So, we need to have them both merged into one log stream when capturing. I don't think we can do this post run, right? :thinking:

sebastian-quintero replied:
Fixed to stream stdout and stderr to the log file for multi-file

merschformann commented on nextmv/nextmv/local/executor.py:
```diff
@@ -0,0 +1,565 @@
+        The stdout output of the run, parsed as a dictionary.
+    """
+
+    stats_src = os.path.join(temp_run_outputs_dir, STATISTICS_KEY)
```
I think in case of `multi-file` statistics are just one file (which location can be configured now - live in prod). So, I don't think a directory works here, right?
For `csv-archive` and `json` just checking the `statistics` key makes sense. I am not sure whether ignoring the content type of the run is stable enough, but might be fine.

sebastian-quintero replied:
Adapted for multi-file to use the configurations in the manifest if they are defined or defaulting to this when not.

merschformann commented on nextmv/nextmv/local/geojson_handler.py:
```diff
@@ -0,0 +1,323 @@
+            return tooltip_fields, popup_fields
+
+        available_fields = list(first_feature["properties"].keys())
+        # Prioritize common field names for tooltip/popup
+        priority_fields = ["name", "title", "label", "id", "popupContent", "description"]
+
+        for field in priority_fields:
+            if field in available_fields:
+                tooltip_fields.append(field)
```
I like the addition of tooltips and popups, but I think we should align with what Console does. I think, Console uses specific fields/keys for different intentions:

- "properties.name": Used for filtering (and the popup title?)
- "properties.style": A _map_ that understands "color", "fillColor", etc.
- "properties.metadata": A _list_ of key-value pairs like these: `{"key": "Vehicle ID", "value": "v1"}`

We should sync on this again though. I have an example of this, I think and we can check with David.

merschformann commented on nextmv/nextmv/local/runner.py:
```diff
@@ -0,0 +1,312 @@
+        stdout=subprocess.DEVNULL,
+        stderr=subprocess.DEVNULL,
+        cwd=os.path.dirname(__file__),
+        start_new_session=True,  # Detach from parent process
```
Is there a way to interact with an ongoing run(s)? Maybe not right now, but in future? :thinking: 
I can easily imagine spawning a number of runs (maybe even accidentally) and then not knowing when they are done, what their status is, etc.

sebastian-quintero replied:
Yes, when a run is ongoing you can poll for results and the normal stuff.

merschformann commented on nextmv/tests/local/__init__.py:
Should this have explicit exports? :thinking:

sebastian-quintero replied:
I avoided it to prevent importing optional deps.

merschformann commented on nextmv/nextmv/run.py:
Will the types of run be dual exported (from cloud AND from local)? :thinking: 
I hope this is all not causing too much confusion later. Now that we are adding yet another role for `nextmv` as a package.

sebastian-quintero replied:
I am still exporting stuff from `cloud` for backwards compatibility but in the future the types established as part of `nextmv` should only be exported from `nextmv`.

merschformann reviewed (approved):
I am running out of review time. :sweat_smile: 

Looks good. I do have some thoughts in general and left some comments. I think it would be good to release small things first before unleashing this kraken. :blush: 
I don't want to block this though. Happy to sync internally. 

Great work! Very clean! :hugs:

sebastian-quintero committed:
Merge branch 'develop' into feature/eng-6317-add-ability-to-execute-a-local-run

sebastian-quintero committed:
Refactors input validation

Validates input directory path
and configuration for local and
cloud applications.

sebastian-quintero committed:
Refactors temp directory usage

Simplifies temp directory creation.

Removes redundant validations.

sebastian-quintero committed:
Adds multi-file support

Adds support for custom file locations.

sebastian-quintero committed:
Handles run errors correctly

Ensures proper error logging.

Fixes output processing.

sebastian-quintero committed:
Ignores .nextmv directory

Prevents tracking of
the local .nextmv
directory.

sebastian-quintero committed:
Updates local executor tests

Refactors tests for manifest.
Ensures correct input processing.
Improves test coverage.

sebastian-quintero committed:
Handles custom file paths

Adds support for custom
paths for assets and
statistics files in
manifests.
